byzcoin/statetrie_test.go

```diff
@@ -71,3 +82,125 @@ func TestStateTrie(t *testing.T) {
 	require.NoError(t, sst.Commit())
 	require.True(t, bytes.Equal(sst.GetRoot(), newRoot))
 }
+
+// TestDarcRetrieval is used as a benchmark here. It stores a lot of darcs
+// that are all linked in a tree-structure, and then uses EvalExprDarc to
+// check whether an identity is valid or not.
+// Use with
+//   go test -v -cpuprofile cpu.prof -run TestDarcRetrieval
+//   go tool pprof -pdf cpu.prof
+func TestDarcRetrieval(t *testing.T) {
+	darcDepth := 3
+	darcWidth := 10
+	entries := 10000
+
+	s := newSer(t, 1, 10*time.Second)
+	defer s.local.CloseAll()
+	s.local.Check = onet.CheckNone
+
+	st, err := s.service().getStateTrie(s.genesis.SkipChainID())
+	require.NoError(t, err)
+	require.NotNil(t, st)
+	require.NotEqual(t, -1, st.GetIndex())
+
+	log.Lvl1("Creating random entries")
+	value := random.Bits(1024, true, random.New())
+	start := time.Now()
+	var scs []StateChange
+	for e := 1; e <= entries; e++ {
+		scs = append(scs, StateChange{
+			StateAction: Create,
+			InstanceID:  random.Bits(256, true, random.New()),
+			ContractID:  ContractDarcID,
+			Value:       value,
+		})
+		if e%100 == 0 {
+			st.StoreAll(scs, 5, CurrentVersion)
+			scs = []StateChange{}
+			log.Print(e, time.Now().Sub(start))
+			start = time.Now()
+		}
+	}
+
+	log.Lvl1("Creating darcs")
+	counter := big.NewInt(0)
+	// Store the tree of darcs in a 2D-array with the root in index 0 and
+	// the leafs in index darcDepth-1
+	darcs := make([][]*darc.Darc, darcDepth)
+	for d := darcDepth - 1; d >= 0; d-- {
+		width := math.Exp(big.NewInt(int64(darcWidth)),
+			big.NewInt(int64(d))).Int64()
+		darcs[d] = make([]*darc.Darc, width)
+		// The leafs just hold a dummy reference to a darc
+		if d == darcDepth-1 {
```

Comment:
This a "fire once" `if`, move it before the loop.

Reply:
but then I need to calculate the `width` and assign the `darcs[d]` twice - once before the loop and once in the loop.

Reply:
you can use an helper function if wanted
```golang
getDarcSliceForDepth := func(depth uint) []*darc.Darc
```

Reply:
That's more complicated and unreadable for me.